### ISSUE TITLE
Work with users' primary Google calendars

### ISSUE THREAD

**Issue by timmy-mac** [jamesgurung/timetable-calendar-generator#4]
I might be stretching the bounds of reality now, in terms of my ability to drag a school into the 21st Century, and I'm certainly not expecting you to write the code within minutes, as per my previous question!

This has been playing over in my head a bit:
> A: "I can do Tuesday period 4"
> B: "I've got year 11, how about Wednesday 2?"
> A: "I can do Wednesday 1?!"
>...

I've just been playing around with Google Calendar a bit and it seems that it will only pull 'Busy' information from a user's primary calendar (even though we have the option to both input free/busy for an event, and to share free/busy in general within the calendar settings?!?). As you can see here, it's not even aware that I'm 'busy' during the events in _my_ timetable!
![image](https://user-images.githubusercontent.com/34206922/62971530-fbccd900-be09-11e9-974f-9c9d535acd2e.png)

When I first started using your application, I was pleased that it wasn't messing around with anyone's primary calendar. Now I'm thinking I might quite like it to.

As before, I'm prepared to get up and running and have a crack myself. I just wondered what you think the potential bear traps might be?

**Comment by jamesgurung** [comment]
I agree, I always had this frustration with Google Calendar's interpretation of free/busy times.

The problem is one of **risk management** - at the moment, if there's an bug in the code then it will only affect the My Timetable calendar. As soon as you start editing users' primary calendars, there's the risk that an unexpected error causes their own events to be changed or deleted. Bearing in mind that this risk applies to all colleagues, it is arguably not worth the small benefit of easier scheduling.

I'd certainly consider a pull request if you're interested to play around with this (at your own risk!)

1. Make a copy of [GoogleCalendarWriter.cs](https://github.com/jamesgurung/timetable-calendar-generator/blob/master/src/CalendarWriters/Google/GoogleCalendarWriter.cs) and call it `GooglePrimaryCalendarWriter`. Then play around in there.
1. You'll need to make trivial changes to OutputType, ArgumentParser, and CalendarWriterFactory to support this new output type. It could be called with a new argument, e.g. `--googleprimary`.
1. The `GoogleCalendarWriter` is currently set to *delete all future events* which do not match the current calendar - you would definitely need to change this! A possible approach would be to add a tag (e.g. "Auto-generated by makecal") to the description of each event created by the app. Then when filtering events for deletion, only consider ones which contain this tag.
1. For the timetable to remain a different colour, you'd need to set this manually for each event.

**Comment by timmy-mac** [comment]
> 3. The `GoogleCalendarWriter` is currently set to _delete all future events_ which do not match the current calendar - you would definitely need to change this! A possible approach would be to add a tag (e.g. "Auto-generated by makecal") to the description of each event created by the app. Then when filtering events for deletion, only consider ones which contain this tag.

That makes sense.. Am I right in I edit something in GooglePrimaryCalendarWriter to get the "generated by makecal" tag in there but that it might make sense to play around in the EventComparer to look for it and mark for deletion?

**Comment by jamesgurung** [comment]
I think it's easier to edit the equivalent of [this line](https://github.com/jamesgurung/timetable-calendar-generator/blob/d94fda2fbdf967c1a48372b80595155144c36e50/src/CalendarWriters/Google/GoogleCalendarWriter.cs#L43):

```c#
existingEvents.Where(o => o.Description == "...").Except(expectedEvents, comparer)

**Comment by jamesgurung** [comment]
You'll also need to make sure the Description field (I'm not sure if that's what it's called) is requested on [this line](https://github.com/jamesgurung/timetable-calendar-generator/blob/d94fda2fbdf967c1a48372b80595155144c36e50/src/CalendarWriters/Google/GoogleCalendarWriter.cs#L83), because currently it's not. There may even be a way to filter the original list request, rather than requesting all events and filtering them on the client - I'm not sure if this is possible.

**Comment by jamesgurung** [comment]
I've just taken a quick look at the [events.list API docs](https://developers.google.com/calendar/v3/reference/events/list), and wonder whether it would be better to use a "private extended property" instead of the description tag. This property (e.g. `timetable=true`) could be set whenever an event is created, and used as a filter when listing existing events. This way, non-timetable events would never even be returned by the API.

Also note that you can refer to the primary calendar with the string "primary", so there's no need for `GetCalendarIdAsync()`.

**Comment by timmy-mac** [comment]
Great idea! I'll see what I can do now.  

As predicted, I've taken a bit of time just to get a dev environment set up. I *think* I'm now ready to have a play (having installed the dotnet core SDK in a Linux container on a Chromebook). Had to temporarily go back to .Net 2.2 though.

**Comment by timmy-mac** [comment]
...and I'm struggling with including the external packages and informative errors like "Operation not supported on this platform".  I'm one of those strange creatures that doesn't keep a Windows machine around. I can remote into one though...  will see if I can get SDK installed.

I hadn't understood that .Net Framework and .Net Core were slightly different beasts anyway.

Tim

**Comment by jamesgurung** [comment]
You'll need the [.NET Core 3 SDK](https://dotnet.microsoft.com/download/dotnet-core/3.0), and then you can call `dotnet run` from the command line. It should just work. I recommend [VS Code](https://code.visualstudio.com/) as an editor.

**Comment by timmy-mac** [comment]
Thanks again James.  I've got this working now, using the PrivateExtendedProperty.  

Will test a little more before I make pull request for you to consider..!

I think you're right hat a colour change would be nice.  Will see if I can figure that out too.

**Comment by jamesgurung** [comment]
Thanks Tim, I think this will be a great contribution to the project.

**Comment by jamesgurung** [comment]
I've made a small change to simplify the command line arguments:

    --google
    --google --primary
    --google --remove-secondary

I've also made some other tweaks to the code. However, I no longer work in a school that uses Google Calendar, so I have no way of testing it!

Is there any chance you could try it out (on a test user), please?

**Comment by timmy-mac** [comment]
Nice work! Makes much more sense ..

Yes, will test. Sorry to hear about your lack of gsuite!!

**Comment by timmy-mac** [comment]
> (1/1) t.mc*@**.org               Failed. Object reference not set to an instance of an object.

Getting the above when I run with '--google --remove-secondary'

I have run with just '--google' first, so 'My timetable' exists. Finding the debug tool in VS code doesn't tell me much, or I don't know how to use it properly.

I think this is unrelated but wondered if you knew about this too:
![Screenshot 2019-08-19 at 07 53 42](https://user-images.githubusercontent.com/34206922/63268941-36f95d00-c28d-11e9-94a8-912d18e24097.png)

--ical and --csv both seem to work though.

**Comment by jamesgurung** [comment]
Thanks for trying it out.

That's a really frustrating error message, and I have no idea what it's referring to! Please could you try commenting out both `try/catch` blocks in `Program.cs` to see if we can narrow down the issue? Run it with the debugger in vscode. You can set command line args by clicking the cog icon.

![VS Code debugger](https://docs.microsoft.com/en-us/dotnet/core/tutorials/media/with-visual-studio-code/select-net-core.png)

As you say, I think the switch statement is unrelated, and it might just be that we're running different versions of the editor. I tried casting both expressions to `ICalendarWriter` but that's showing up as redundant. I guess if it compiles then it's not an issue.

![Screenshot](https://user-images.githubusercontent.com/1919846/63286174-4f7a6f00-c2af-11e9-8b59-2a1db9aaf1b2.png)

**Comment by jamesgurung** [comment]
Ok I think I've found the issue and fixed in d1023d517bfe691e415f9c75cb03f874ae54eab0. The service account key wasn't being picked up correctly. Please can you try the latest version?

Please also check that `--google --primary` still works as expected, i.e. leaves user-created events alone.

If it works then I'll publish a release. :+1:

**Comment by timmy-mac** [comment]
Good new - this is now all working brilliantly.  I'm very happy with the changes indeed.

If I'm being picky, I'd suggest changing "Calendar generation complete" to "Operation complete", just so it makes sense when you're running with the --remove flag.  ...or bothering to make it say something different in each case. 

One for another time, perhaps..   results approach!

**Comment by jamesgurung** [comment]
Great, thanks Tim. This feature is now published in the [v4.0 release](https://github.com/jamesgurung/timetable-calendar-generator/releases/tag/v4.0)!
